Examples/Algorithms/MaterialMapping/src/GeantinoRecording.cpp

```diff
@@ -0,0 +1,77 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2017-2018 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ACTFW/MaterialMapping/GeantinoRecording.hpp"
+
+#include <FTFP_BERT.hh>
+#include <iostream>
+#include <stdexcept>
+
+#include "ACTFW/Framework/WhiteBoard.hpp"
+#include "ACTFW/Plugins/Geant4/MMDetectorConstruction.hpp"
+#include "ACTFW/Plugins/Geant4/MMEventAction.hpp"
+#include "ACTFW/Plugins/Geant4/MMPrimaryGeneratorAction.hpp"
+#include "ACTFW/Plugins/Geant4/MMRunAction.hpp"
+#include "ACTFW/Plugins/Geant4/MMSteppingAction.hpp"
+
+FW::GeantinoRecording::GeantinoRecording(
+    const FW::GeantinoRecording::Config& cnf, Acts::Logging::Level level)
+    : FW::BareAlgorithm("GeantinoRecording", level),
+      m_cfg(cnf),
+      m_runManager(std::make_unique<G4RunManager>()) {
+  /// Check if the geometry should be accessed over the geant4 service
+  if (m_cfg.geant4Service) {
+    m_runManager->SetUserInitialization(m_cfg.geant4Service->geant4Geometry());
+  } else if (!m_cfg.gdmlFile.empty()) {
+    /// Access the geometry from the gdml file
+    ACTS_INFO(
+        "received Geant4 geometry from GDML file: " << m_cfg.gdmlFile.c_str());
+    FW::Geant4::MMDetectorConstruction* detConstruction =
+        new FW::Geant4::MMDetectorConstruction();
+    detConstruction->setGdmlInput(m_cfg.gdmlFile.c_str());
+    m_runManager->SetUserInitialization(
+        detConstruction);  // constructs detector (calls Construct in
+                           // Geant4DetectorConstruction)
+  } else {
+    throw std::invalid_argument("Missing geometry input for Geant4");
+  }
+
+  /// Now set up the Geant4 simulation
+  m_runManager->SetUserInitialization(new FTFP_BERT);
+  m_runManager->SetUserAction(new FW::Geant4::MMPrimaryGeneratorAction(
+      "mu-", 1000., m_cfg.seed1, m_cfg.seed2));
```

Comment:
Going through the material mapping I noticed that in the GeantinoRecording Algorithm the geantino have been replace with muons. I guess you didn't do that and it was done in a previous MR that I didn't look at but could you fix it ?

Reply:
It is also extremely slow compare to what it use to be. I will investigate

Reply:
This was changed in a PR by Stany to be able to do Geant4/Fatras comparison. I can switch this back for now. I the long run, we need to be able to run Geant4 with arbitrary particle input.

Reply:
Isn't that the purpose of the ActsGenParticleGun Algorithm ?

I also notice that in that PR she added the recording of track step which make the reording slower by a factor 60, maybe all that should be in another algorithm (or at least have an option to turn it off) so that the geantino recording only does the geantino recording.

I will remove that part of the code locally for now and continue to try the mapping chain.

Reply:
> Isn't that the purpose of the ActsGenParticleGun Algorithm ?

Yes, but we need to adapt the Geant4 algorithms to take the generated particles as inputs; instead of relying on the Geant4 internal particle gun.

> I also notice that in that PR she added the recording of track step which make the reording slower by a factor 60, maybe all that should be in another algorithm (or at least have an option to turn it off) so that the geantino recording only does the geantino recording.
> 
> I will remove that part of the code locally for now and continue to try the mapping chain.

In that case, a separate algorithm is probably the better choice. I will open an issue.

Reply:
Tracked in #111 now.

Reply:
Thanks !